README.md
# ezcrop

Javascript image crop tool with zooming and rotating.

Ezcrop is a Vanilla JavaScript version of jQuery [Cropit](https://github.com/scottcheng/cropit) plugin by Scott Cheng.


## Features

* Selecting images from filesystem
* Zooming
* Moving
* Rotating
* Cropping
* Export to data URI


## Installation

```bash
# Install ezcrop with bower
$ bower install ezcrop

# or with npm
$ npm install ezcrop

# or just download [this file](dist/ezcrop.js) and use it as you wish
```


## Usage

Add `ezcrop.js` to the page.

```html
<script src="dist/ezcrop.js"></script>
```

And the required bit of HTML:

```html
<div id="image-cropper">
    <div class="ezcrop-preview"></div>
    <input type="file" class="ezcrop-image-input">
    <div class="image-size-label">
      Resize image
    </div>
    <input type="range" class="ezcrop-image-zoom-input">
    <button id="rotate-ccw" class="rotate-ccw">Rotate counterclockwise</button>
    <button id="rotate-cw" class="rotate-cw">Rotate clockwise</button>

    <button id="export" class="export">Export</button>
</div>
```
Initialize

```javascript
<script>
  (function() {
    var cropper = new ezcrop(document.getElementById('image-cropper'), {
      imageState: {
        src: 'images/hands.jpg', // start with a default image
      },
    });

    document.getElementById('rotate-ccw').addEventListener('click', function() {
      cropper.rotateCCW();
    });

    document.getElementById('rotate-cw').addEventListener('click', function() {
      cropper.rotateCW();
    });

    document.getElementById('export').addEventListener('click', function() {
      var imageData = cropper.getCroppedImageData();
      window.open(imageData);
    });
  })();
</script>
```


## Options

Following options are supported:

| Option | Type | Description | Default |
|----------------------------|-----------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------|
| preview | object | The HTML element that displays image preview. |  |
| fileInput | object | File input element. |  |
| zoomSlider | object | Range input element that controls image zoom. |  |
| width | number | Width of image preview in pixels. If set, it will override the CSS property. | null |
| height | number | Height of image preview in pixels. If set, it will override the CSS property. | null |
| imageBackground | boolean | Whether or not to display the background image beyond the preview area. | false |
| imageBackgroundBorderWidth | array or number | Width of background image border in pixels.,The four array elements specify the width of background image width on the top, right, bottom, left side respectively.,The background image beyond the width will be hidden.,If specified as a number, border with uniform width on all sides will be applied. | [0, 0, 0, 0] |
| exportZoom | number | The ratio between the desired image size to export and the preview size.,For example, if the preview size is `300px * 200px`, and `exportZoom = 2`, then,the exported image size will be `600px * 400px`.,This also affects the maximum zoom level, since the exported image cannot be zoomed to larger than its original size. | 1 |
| exportWidth | number | Fixed width of exported image in pixels. This is useful for calculating of dynamic `exportZoom` for responsive designs. | null |
| allowDragNDrop | boolean | When set to true, you can load an image by dragging it from local file browser onto the preview area. | true |
| minZoom | string | This options decides the minimal zoom level of the image.,If set to `'fill'`, the image has to fill the preview area, i.e. both width and height must not go smaller than the preview area.,If set to `'fit'`, the image can shrink further to fit the preview area, i.e. at least one of its edges must not go smaller than the preview area. | fill |
| maxZoom | number | Determines how big the image can be zoomed. E.g. if set to 1.5, the image can be zoomed to 150% of its original size. | 1 |
| initialZoom | string | `Determines the zoom when an image is loaded.,When set to `'min'`, image is zoomed to the smallest when loaded.,When set to `'image'`, image is zoomed to 100% when loaded. | min |
| freeMove | boolean | When set to true, you can freely move the image instead of being bound to the container borders. | false |
| smallImage | string | When set to `'reject'`, `onImageError` would be called when ezcrop loads an image that is smaller than the container.,When set to `'allow'`, images smaller than the container can be zoomed down to its original size, overiding `minZoom` option.,When set to `'stretch'`, the minimum zoom of small images would follow `minZoom` option. | reject |


## Callbacks

Following callbacks can be used to interact with `ezcrop`.

| Callback | Description |
|------------------------|----------------------------------------------------------------------------|
| onFileChange(event) | Called when user selects a file in the select file input. |
| onFileReaderError | Called when `FileReader` encounters an error while loading the image file. |
| onImageLoading | Called when image starts to be loaded. |
| onImageLoaded | Called when image is loaded. |
| onImageError(error) | Called when image cannot be loaded. |
| onZoomEnabled | Called when image the zoom slider is enabled. |
| onZoomDisabled | Called when image the zoom slider is disabled. |
| onZoomChange(zoom) | Called when zoom changes. |
| onOffsetChange(offset) | Called when image offset changes. |


## Development

* Build: `npm run build`
  * Watch for changes and rebuild: `webpack -w`
* Test: `npm test`
  * Test specific file: `jest <filename>`
* Lint: `npm run jshint -s`


## License

MIT


CHANGELOG.md
## 0.0.3 (October 21, 2017) [by noo-zh](https://github.com/noo-zh/ezcrop)

### Features

* Dynamic `exportZoom` using `exportWidth`
* Production bundle optimization

## 0.0.2 (?)

Not documented.

## 0.0.1 (December 26, 2016)

### Features

* Selecting images from filesystem
* Zooming
* Moving
* Rotating
* Cropping
* Export to data URI

src/ezcrop.js
import Zoomer from './zoomer';
import { CLASS_NAMES, ERRORS, EVENTS } from './constants';
import { loadDefaults } from './options';
import { exists, round } from './utils';

class Ezcrop {
  constructor(element, options) {
    this.el = element;

    const defaults = loadDefaults(this.el);
    this.options = extend({}, defaults, options);

    this.init();
  }

  init() {
    this.img = new Image();
    this.preImage = new Image();
    this.img.onload = this.onImageLoaded.bind(this);
    this.preImage.onload = this.onPreImageLoaded.bind(this);
    this.img.onerror = this.preImage.onerror = () => {
      this.onImageError.call(this, ERRORS.IMAGE_FAILED_TO_LOAD);
    };

    this.preview = this.options.preview;
    this.fileInput = this.options.fileInput;
    this.zoomSlider = this.options.zoomSlider;

    this.preview.style.position = 'relative';
    this.fileInput.setAttribute('accept', 'image/*');
    this.zoomSlider.setAttribute('min', 0);
    this.zoomSlider.setAttribute('max', 1);
    this.zoomSlider.setAttribute('step', 0.01);

    this.previewSize = {
      width: this.options.width || this.preview.clientWidth,
      height: this.options.height || this.preview.clientHeight,
    };

    this.image = document.createElement('img');
    addClass(this.image, CLASS_NAMES.PREVIEW_IMAGE)
    this.image.setAttribute('alt', '');
    this.image.style.transformOrigin = 'top left';
    this.image.style.webkitTransformOrigin = 'top left';

    this.imageContainer = document.createElement('div');
    addClass(this.imageContainer, CLASS_NAMES.PREVIEW_IMAGE_CONTAINER)
    this.imageContainer.setAttribute('style', 'position: absolute; overflow: hidden; left: 0; top: 0; width: 100%; height: 100%;');
    this.imageContainer.appendChild(this.image);
    this.preview.appendChild(this.imageContainer);

    if (this.options.imageBackground) {
      if (Array.isArray(this.options.imageBackgroundBorderWidth)) {
        this.bgBorderWidthArray = this.options.imageBackgroundBorderWidth;
      } else {
        this.bgBorderWidthArray = [0, 1, 2, 3].map(() => this.options.imageBackgroundBorderWidth);
      }

      this.bg = document.createElement('img');
      addClass(this.bg, CLASS_NAMES.PREVIEW_BACKGROUND);
      this.bg.setAttribute('alt', '');
      this.bg.setAttribute('style', 'position: relative; left: ' + this.bgBorderWidthArray[3] + '; top: ' + this.bgBorderWidthArray[0] + '; transform-origin: top left; -webkit-transform-origin: top left;');

      this.bgContainer = document.createElement('div');
      addClass(this.bgContainer, CLASS_NAMES.PREVIEW_BACKGROUND_CONTAINER);
      this.bgContainer.setAttribute('style', 'position: absolute; z-index: 0; top: -' + this.bgBorderWidthArray[0] + '; right: -' + this.bgBorderWidthArray[1] + '; bottom: -' + this.bgBorderWidthArray[2] + '; left: -' + this.bgBorderWidthArray[3] + ';');
      this.bgContainer.appendChild(this.bg);
      if (this.bgBorderWidthArray[0] > 0) {
        this.bgContainer.style.overflow = 'hidden';
      }
      this.preview.insertBefore(this.bgContainer, this.preview.firstChild);
    }

    this.initialZoom = this.options.initialZoom;

    this.imageLoaded = false;

    this.moveContinue = false;

    this.zoomer = new Zoomer();

    if (this.options.allowDragNDrop) {
      // $.event.props.push('dataTransfer');
    }

    this.bindListeners();

    if (this.options.imageState && this.options.imageState.src) {
      this.loadImage(this.options.imageState.src);
    }

    this.el.setAttribute('data-ezcrop', this);
  }

  bindListeners() {
    on(this.fileInput, ['change'], this.onFileChange.bind(this));
    on(this.imageContainer, EVENTS.PREVIEW, this.onPreviewEvent.bind(this));
    on(this.zoomSlider, EVENTS.ZOOM_INPUT, this.onZoomSliderChange.bind(this));

    if (this.options.allowDragNDrop) {
      on(this.imageContainer, ['dragover', 'dragleave'], this.onDragOver.bind(this));
      on(this.imageContainer, ['drop'], this.onDrop.bind(this));
    }
  }

  unbindListeners() {
    off(this.fileInput, ['change'], this.onFileChange);
    off(this.imageContainer, EVENTS.PREVIEW, this.onPreviewEvent);
    off(this.imageContainer, ['dragover', 'dragleave', 'drop'], this.onDragOver);
    off(this.zoomSlider, EVENTS.ZOOM_INPUT, this.onZoomSliderChange);
  }

  onFileChange(e) {
    this.options.onFileChange(e);

    if (this.fileInput.files) {
      this.loadFile(this.fileInput.files[0]);
    }
  }
  
  calculateExportZoom() {
    let exportZoom = this.options.exportZoom;
    
    if (this.options.exportWidth && this.options.exportWidth >= this.previewSize.width) {
      exportZoom = this.options.exportWidth / this.previewSize.width;
    }
    
    return exportZoom;
  }

  loadFile(file) {
    const fileReader = new FileReader();
    if (file && file.type.match('image')) {
      fileReader.readAsDataURL(file);
      fileReader.onload = this.onFileReaderLoaded.bind(this);
      fileReader.onerror = this.onFileReaderError.bind(this);
    } else if (file) {
      this.onFileReaderError();
    }
  }

  onFileReaderLoaded(e) {
    this.loadImage(e.target.result);
  }

  onFileReaderError() {
    this.options.onFileReaderError();
  }

  onDragOver(e) {
    e.preventDefault();
    e.dataTransfer.dropEffect = 'copy';
    if(e.type === 'dragover') {
      addClass(this.preview, CLASS_NAMES.DRAG_HOVERED);
    } else {
      removeClass(this.preview, CLASS_NAMES.DRAG_HOVERED);
    }
  }

  onDrop(e) {
    e.preventDefault();
    e.stopPropagation();

    const files = Array.prototype.slice.call(e.dataTransfer.files, 0);
    files.some((file) => {
      if (!file.type.match('image')) { return false; }

      this.loadFile(file);
      return true;
    });

    // this.preview.classList.remove(CLASS_NAMES.DRAG_HOVERED);
    removeClass(this.preview, CLASS_NAMES.DRAG_HOVERED);
  }

  loadImage(imageSrc) {
    if (!imageSrc) { return; }

    this.options.onImageLoading();
    this.setImageLoadingClass();

    if (imageSrc.indexOf('data') === 0) {
      this.preImage.src = imageSrc;
    } else {
      const xhr = new XMLHttpRequest();
      xhr.onload = (e) => {
        if (e.target.status >= 300) {
          this.onImageError.call(this, ERRORS.IMAGE_FAILED_TO_LOAD);
          return;
        }

        this.loadFile(e.target.response);
      };
      xhr.open('GET', imageSrc);
      xhr.responseType = 'blob';
      xhr.send();
    }
  }

  onPreImageLoaded() {
    if (this.shouldRejectImage({
      imageWidth: this.preImage.width,
      imageHeight: this.preImage.height,
      previewSize: this.previewSize,
      maxZoom: this.options.maxZoom,
      exportZoom: this.calculateExportZoom(),
      smallImage: this.options.smallImage,
    })) {
      this.onImageError(ERRORS.SMALL_IMAGE);
      if (this.img.src) { this.setImageLoadedClass(); }
      return;
    }

    this.img.src = this.preImage.src;
  }

  onImageLoaded() {
    this.rotation = 0;
    this.setupZoomer(this.options.imageState && this.options.imageState.zoom || this._initialZoom);
    if (this.options.imageState && this.options.imageState.offset) {
      this.offset = this.options.imageState.offset;
    } else {
      this.centerImage();
    }

    this.options.imageState = {};

    this.image.setAttribute('src', this.img.src);
    if (this.options.imageBackground) {
      this.bg.setAttribute('src', this.img.src);
    }

    this.setImageLoadedClass();

    this.imageLoaded = true;

    this.options.onImageLoaded();
  }

  onImageError() {
    this.options.onImageError.apply(this, arguments);
    this.removeImageLoadingClass();
  }

  setImageLoadingClass() {
    removeClass(this.preview, CLASS_NAMES.IMAGE_LOADED);
    addClass(this.preview, CLASS_NAMES.IMAGE_LOADING);
  }

  setImageLoadedClass() {
    removeClass(this.preview, CLASS_NAMES.IMAGE_LOADING);
    addClass(this.preview, CLASS_NAMES.IMAGE_LOADED);
  }

  removeImageLoadingClass() {
    removeClass(this.preview, CLASS_NAMES.IMAGE_LOADING);
  }

  getEventPosition(e) {
    if (e.originalEvent && e.originalEvent.touches && e.originalEvent.touches[0]) {
      e = e.originalEvent.touches[0];
    }
    if (e.clientX && e.clientY) {
      return { x: e.clientX, y: e.clientY };
    }
  }

  onPreviewEvent(e) {
    if (!this.imageLoaded) { return; }

    this.moveContinue = false;
    off(this.imageContainer, EVENTS.PREVIEW_MOVE, this.onMove);

    if (e.type === 'mousedown' || e.type === 'touchstart') {
      e.preventDefault();
      this.origin = this.getEventPosition(e);
      this.moveContinue = true;
      on(this.imageContainer, EVENTS.PREVIEW_MOVE, this.onMove.bind(this));
    } else {
      document.body.focus();
    }

    e.stopPropagation();
    return false;
  }

  onMove(e) {
    const eventPosition = this.getEventPosition(e);

    if (this.moveContinue && eventPosition) {
      this.offset = {
        x: this.offset.x + eventPosition.x - this.origin.x,
        y: this.offset.y + eventPosition.y - this.origin.y,
      };
    }

    this.origin = eventPosition;

    e.stopPropagation();
    return false;
  }

  set offset(position) {
    if (!position || !exists(position.x) || !exists(position.y)) { return; }

    this._offset = this.fixOffset(position);
    this.renderImage();

    this.options.onOffsetChange(position);
  }

  fixOffset(offset) {
    if (!this.imageLoaded) { return offset; }

    const ret = { x: offset.x, y: offset.y };

    if (!this.options.freeMove) {
      if (this.imageWidth * this.zoom >= this.previewSize.width) {
        ret.x = Math.min(0, Math.max(ret.x,
          this.previewSize.width - this.imageWidth * this.zoom));
      }
      else {
        ret.x = Math.max(0, Math.min(ret.x,
          this.previewSize.width - this.imageWidth * this.zoom));
      }

      if (this.imageHeight * this.zoom >= this.previewSize.height) {
        ret.y = Math.min(0, Math.max(ret.y,
          this.previewSize.height - this.imageHeight * this.zoom));
      }
      else {
        ret.y = Math.max(0, Math.min(ret.y,
          this.previewSize.height - this.imageHeight * this.zoom));
      }
    }

    ret.x = round(ret.x);
    ret.y = round(ret.y);

    return ret;
  }

  centerImage() {
    if (!this.img.width || !this.img.height || !this.zoom) { return; }

    this.offset = {
      x: (this.previewSize.width - this.imageWidth * this.zoom) / 2,
      y: (this.previewSize.height - this.imageHeight * this.zoom) / 2,
    };
  }

  onZoomSliderChange() {
    if (!this.imageLoaded) { return; }

    this.zoomSliderPos = Number(this.zoomSlider.value);
    const newZoom = this.zoomer.getZoom(this.zoomSliderPos);
    if (newZoom === this.zoom) { return; }
    this.zoom = newZoom;
  }

  enableZoomSlider() {
    this.zoomSlider.removeAttribute('disabled');
    this.options.onZoomEnabled();
  }

  disableZoomSlider() {
    this.zoomSlider.setAttribute('disabled', true);
    this.options.onZoomDisabled();
  }

  setupZoomer(zoom) {
    this.zoomer.setup({
      imageSize: this.imageSize,
      previewSize: this.previewSize,
      exportZoom: this.calculateExportZoom(),
      maxZoom: this.options.maxZoom,
      minZoom: this.options.minZoom,
      smallImage: this.options.smallImage,
    });
    this.zoom = exists(zoom) ? zoom : this._zoom;

    if (this.isZoomable()) {
      this.enableZoomSlider();
    }
    else {
      this.disableZoomSlider();
    }
  }

  set zoom(newZoom) {
    newZoom = this.fixZoom(newZoom);

    if (this.imageLoaded) {
      const oldZoom = this.zoom;

      const newX = this.previewSize.width / 2 - (this.previewSize.width / 2 - this.offset.x) * newZoom / oldZoom;
      const newY = this.previewSize.height / 2 - (this.previewSize.height / 2 - this.offset.y) * newZoom / oldZoom;

      this._zoom = newZoom;
      this.offset = { x: newX, y: newY }; // Triggers renderImage()
    }
    else {
      this._zoom = newZoom;
    }

    this.zoomSliderPos = this.zoomer.getSliderPos(this.zoom);
    this.zoomSlider.value = this.zoomSliderPos;

    this.options.onZoomChange(newZoom);
  }

  fixZoom(zoom) {
    return this.zoomer.fixZoom(zoom);
  }

  isZoomable() {
    return this.zoomer.isZoomable();
  }

  get rotatedOffset() {
    return {
      x: this.offset.x +
          (this.rotation === 90 ? this.img.height * this.zoom : 0) +
          (this.rotation === 180 ? this.img.width * this.zoom : 0),
      y: this.offset.y +
          (this.rotation === 180 ? this.img.height * this.zoom : 0) +
          (this.rotation === 270 ? this.img.width * this.zoom : 0),
    };
  }

  renderImage() {
    const transformation = `
      translate(${this.rotatedOffset.x}px, ${this.rotatedOffset.y}px)
      scale(${this.zoom})
      rotate(${this.rotation}deg)`;

    this.image.style.transform = transformation;
    this.image.style.webkitTransform = transformation;

    if (this.options.imageBackground) {
      this.bg.style.transform = transformation;
      this.bg.style.webkitTransform = transformation;
    }
  }

  set rotation(newRotation) {
    this._rotation = newRotation;

    if (this.imageLoaded) {
      // Change in image size may lead to change in zoom range
      this.setupZoomer();
    }
  }

  get rotation() {
    return this._rotation;
  }

  rotateCW() {
    if (this.shouldRejectImage({
      imageWidth: this.img.height,
      imageHeight: this.img.width,
      previewSize: this.previewSize,
      maxZoom: this.options.maxZoom,
      exportZoom: this.calculateExportZoom(),
      smallImage: this.options.smallImage,
    })) {
      this.rotation = (this.rotation + 180) % 360;
    }
    else {
      this.rotation = (this.rotation + 90) % 360;
    }
  }

  rotateCCW() {
    if (this.shouldRejectImage({
      imageWidth: this.img.height,
      imageHeight: this.img.width,
      previewSize: this.previewSize,
      maxZoom: this.options.maxZoom,
      exportZoom: this.calculateExportZoom(),
      smallImage: this.options.smallImage,
    })) {
      this.rotation = (this.rotation + 180) % 360;
    }
    else {
      this.rotation = (this.rotation + 270) % 360;
    }
  }

  shouldRejectImage({ imageWidth, imageHeight, previewSize, maxZoom, exportZoom, smallImage }) {
    if (smallImage !== 'reject') { return false; }

    return imageWidth * maxZoom < previewSize.width * exportZoom ||
           imageHeight * maxZoom < previewSize.height * exportZoom;
  }

  getCroppedImageData(exportOptions) {
    if (!this.img.src) { return; }

    const exportDefaults = {
      type: 'image/png',
      quality: 0.75,
      originalSize: false,
      fillBg: '#fff',
    };
    exportOptions = extend({}, exportDefaults, exportOptions);

    const exportZoom = exportOptions.originalSize ? 1 / this.zoom : this.calculateExportZoom();

    const zoomedSize = {
      width: this.zoom * exportZoom * this.img.width,
      height: this.zoom * exportZoom * this.img.height,
    };

    const canvas = document.createElement('canvas');
    canvas.setAttribute('width', this.previewSize.width * exportZoom);
    canvas.setAttribute('height', this.previewSize.height * exportZoom);

    const canvasContext = canvas.getContext('2d');

    if (exportOptions.type === 'image/jpeg') {
      canvasContext.fillStyle = exportOptions.fillBg;
      canvasContext.fillRect(0, 0, canvas.width, canvas.height);
    }

    canvasContext.translate(
      this.rotatedOffset.x * exportZoom,
      this.rotatedOffset.y * exportZoom);
    canvasContext.rotate(this.rotation * Math.PI / 180);
    canvasContext.drawImage(this.img,
      0, 0,
      zoomedSize.width,
      zoomedSize.height);

    return canvas.toDataURL(exportOptions.type, exportOptions.quality);
  }

  get imageState() {
    return {
      src: this.img.src,
      offset: this.offset,
      zoom: this.zoom,
    };
  }

  get imageSrc() {
    return this.img.src;
  }

  set imageSrc(imageSrc) {
    this.loadImage(imageSrc);
  }

  get offset() {
    return this._offset;
  }

  get zoom() {
    return this._zoom;
  }

  get imageWidth() {
    return this.rotation % 180 === 0 ? this.img.width : this.img.height;
  }

  get imageHeight() {
    return this.rotation % 180 === 0 ? this.img.height : this.img.width;
  }

  get imageSize() {
    return {
      width: this.imageWidth,
      height: this.imageHeight,
    };
  }

  get initialZoom() {
    return this.options.initialZoom;
  }

  set initialZoom(initialZoomOption) {
    this.options.initialZoom = initialZoomOption;
    if (initialZoomOption === 'min') {
      this._initialZoom = 0; // Will be fixed when image loads
    }
    else if (initialZoomOption === 'image') {
      this._initialZoom = 1;
    }
    else {
      this._initialZoom = 0;
    }
  }

  get exportZoom() {
    return this.options.exportZoom;
  }

  get calculatedExportZoom() {
    return this.calculateExportZoom();
  }

  set exportZoom(exportZoom) {
    this.options.exportZoom = exportZoom;
    this.setupZoomer();
  }

  get minZoom() {
    return this.options.minZoom;
  }

  set minZoom(minZoom) {
    this.options.minZoom = minZoom;
    this.setupZoomer();
  }

  get maxZoom() {
    return this.options.maxZoom;
  }

  set maxZoom(maxZoom) {
    this.options.maxZoom = maxZoom;
    this.setupZoomer();
  }

  get previewSize() {
    return this._previewSize;
  }

  set previewSize(size) {
    if (!size || size.width <= 0 || size.height <= 0) { return; }

    this._previewSize = {
      width: size.width,
      height: size.height,
    };
    this.preview.style.width = this.previewSize.width + 'px';
    this.preview.style.height = this.previewSize.height + 'px';

    if (this.imageLoaded) {
      this.setupZoomer();
    }
  }

  disable() {
    this.unbindListeners();
    this.disableZoomSlider();
    this.el.addClass(CLASS_NAMES.DISABLED);
  }

  reenable() {
    this.bindListeners();
    this.enableZoomSlider();
    this.el.removeClass(CLASS_NAMES.DISABLED);
  }

  $(selector) {
    if (!this.el) { return null; }
    return this.el.querySelectorAll(selector);
  }
}

// Utilities
function extend() {
  for(var i=1; i<arguments.length; i++)
    for(var key in arguments[i])
      if(arguments[i].hasOwnProperty(key))
        arguments[0][key] = arguments[i][key];

  return arguments[0];
}

function addClass(el, className) {
  if (el.classList)
    el.classList.add(className);
  else
    el.className += ' ' + className;
}

function removeClass(el, className) {
  if (el.classList)
    el.classList.remove(className);
  else
    el.className = el.className.replace(new RegExp('(^|\\b)' + className.split(' ').join('|') + '(\\b|$)', 'gi'), ' ');
}

function toggleClass(el, className) {
  if (el.classList) {
    el.classList.toggle(className);
  } else {
    var classes = el.className.split(' ');
    var existingIndex = classes.indexOf(className);

    if (existingIndex >= 0)
      classes.splice(existingIndex, 1);
    else
      classes.push(className);

    el.className = classes.join(' ');
  }
}

function on(el, events, handler) {
  events.forEach((e) => {
    el.addEventListener(e, handler);
  });
}

function off(el, events, handler) {
  events.forEach((e) => {
    el.removeEventListener(e, handler);
  });
}

export default Ezcrop;


src/constants.js
export const PLUGIN_KEY = 'ezcrop';

export const CLASS_NAMES = {
  PREVIEW: 'ezcrop-preview',
  PREVIEW_IMAGE_CONTAINER: 'ezcrop-preview-image-container',
  PREVIEW_IMAGE: 'ezcrop-preview-image',
  PREVIEW_BACKGROUND_CONTAINER: 'ezcrop-preview-background-container',
  PREVIEW_BACKGROUND: 'ezcrop-preview-background',
  FILE_INPUT: 'ezcrop-image-input',
  ZOOM_SLIDER: 'ezcrop-image-zoom-input',

  DRAG_HOVERED: 'ezcrop-drag-hovered',
  IMAGE_LOADING: 'ezcrop-image-loading',
  IMAGE_LOADED: 'ezcrop-image-loaded',
  DISABLED: 'ezcrop-disabled',
};

export const ERRORS = {
  IMAGE_FAILED_TO_LOAD: { code: 0, message: 'Image failed to load.' },
  SMALL_IMAGE: { code: 1, message: 'Image is too small.' },
};

export const EVENTS = {
  PREVIEW: [
    'mousedown', 'mouseup', 'mouseleave',
    'touchstart', 'touchend', 'touchcancel', 'touchleave',
  ],
  PREVIEW_MOVE: ['mousemove', 'touchmove'],
  ZOOM_INPUT: ['mousemove', 'touchmove', 'change'],
};


src/options.js
import { CLASS_NAMES } from './constants';

const options = {
  elements: [
    {
      name: 'preview',
      //description: 'The HTML element that displays image preview.',
      defaultSelector: `.${CLASS_NAMES.PREVIEW}`,
    },
    {
      name: 'fileInput',
      //description: 'File input element.',
      defaultSelector: `input.${CLASS_NAMES.FILE_INPUT}`,
    },
    {
      name: 'zoomSlider',
      //description: 'Range input element that controls image zoom.',
      defaultSelector: `input.${CLASS_NAMES.ZOOM_SLIDER}`,
    },
  ].map((o) => {
    o.type = 'element';
    o.default = `imageCropper.querySelectorAll('${o.defaultSelector}')[0]`;
    return o;
  }),

  values: [
    {
      name: 'width',
      type: 'number',
      //description: 'Width of image preview in pixels. If set, it will override the CSS property.',
      default: null,
    },
    {
      name: 'height',
      type: 'number',
      //description: 'Height of image preview in pixels. If set, it will override the CSS property.',
      default: null,
    },
    {
      name: 'exportWidth',
      type: 'number',
      //description: 'Fixed width of exported image in pixels. This is useful for calculating of dynamic \`exportZoom\` for responsive designs.',
      default: null,
    },
    {
      name: 'imageBackground',
      type: 'boolean',
      //description: 'Whether or not to display the background image beyond the preview area.',
      default: false,
    },
    {
      name: 'imageBackgroundBorderWidth',
      type: 'array or number',
      //description: `Width of background image border in pixels.
      //  The four array elements specify the width of background image width on the top, right, bottom, left side respectively.
      //  The background image beyond the width will be hidden.
      //  If specified as a number, border with uniform width on all sides will be applied.`,
      default: [0, 0, 0, 0],
    },
    {
      name: 'exportZoom',
      type: 'number',
      //description: `The ratio between the desired image size to export and the preview size.
      //  For example, if the preview size is \`300px * 200px\`, and \`exportZoom = 2\`, then
      //  the exported image size will be \`600px * 400px\`.
      //  This also affects the maximum zoom level, since the exported image cannot be zoomed to larger than its original size.`,
      default: 1,
    },
    {
      name: 'allowDragNDrop',
      type: 'boolean',
      //description: 'When set to true, you can load an image by dragging it from local file browser onto the preview area.',
      default: true,
    },
    {
      name: 'minZoom',
      type: 'string',
      //description: `This options decides the minimal zoom level of the image.
      //  If set to \`'fill'\`, the image has to fill the preview area, i.e. both width and height must not go smaller than the preview area.
      //  If set to \`'fit'\`, the image can shrink further to fit the preview area, i.e. at least one of its edges must not go smaller than the preview area.`,
      default: 'fill',
    },
    {
      name: 'maxZoom',
      type: 'number',
      //description: 'Determines how big the image can be zoomed. E.g. if set to 1.5, the image can be zoomed to 150% of its original size.',
      default: 1,
    },
    {
      name: 'initialZoom',
      type: 'string',
      //description: `Determines the zoom when an image is loaded.
      //  When set to \`'min'\`, image is zoomed to the smallest when loaded.
      //  When set to \`'image'\`, image is zoomed to 100% when loaded.`,
      default: 'min',
    },
    {
      name: 'freeMove',
      type: 'boolean',
      //description: 'When set to true, you can freely move the image instead of being bound to the container borders',
      default: false,
    },
    {
      name: 'smallImage',
      type: 'string',
      //description: `When set to \`'reject'\`, \`onImageError\` would be called when ezcrop loads an image that is smaller than the container.
      //  When set to \`'allow'\`, images smaller than the container can be zoomed down to its original size, overiding \`minZoom\` option.
      //  When set to \`'stretch'\`, the minimum zoom of small images would follow \`minZoom\` option.`,
      default: 'reject',
    },
  ],

  callbacks: [
    {
      name: 'onFileChange',
      //description: 'Called when user selects a file in the select file input.',
      params: [
        {
          name: 'event',
          type: 'object',
          //description: 'File change event object',
        },
      ],
    },
    {
      name: 'onFileReaderError',
      //description: 'Called when `FileReader` encounters an error while loading the image file.',
    },
    {
      name: 'onImageLoading',
      //description: 'Called when image starts to be loaded.',
    },
    {
      name: 'onImageLoaded',
      //description: 'Called when image is loaded.',
    },
    {
      name: 'onImageError',
      //description: 'Called when image cannot be loaded.',
      params: [
        {
          name: 'error',
          type: 'object',
          //description: 'Error object.',
        },
        {
          name: 'error.code',
          type: 'number',
          //description: 'Error code. `0` means generic image loading failure. `1` means image is too small.',
        },
        {
          name: 'error.message',
          type: 'string',
          //description: 'A message explaining the error.',
        },
      ],
    },
    {
      name: 'onZoomEnabled',
      //description: 'Called when image the zoom slider is enabled.',
    },
    {
      name: 'onZoomDisabled',
      //description: 'Called when image the zoom slider is disabled.',
    },
    {
      name: 'onZoomChange',
      //description: 'Called when zoom changes.',
      params: [
        {
          name: 'zoom',
          type: 'number',
          //description: 'New zoom.',
        },
      ],
    },
    {
      name: 'onOffsetChange',
      //description: 'Called when image offset changes.',
      params: [
        {
          name: 'offset',
          type: 'object',
          //description: 'New offset, with `x` and `y` values.',
        },
      ],
    },
  ].map((o) => { o.type = 'function'; return o; }),
};

export const loadDefaults = (el) => {
  const defaults = {};
  if (el) {
    options.elements.forEach((o) => {
      defaults[o.name] = el.querySelectorAll(o.defaultSelector)[0];
    });
  }
  options.values.forEach((o) => {
    defaults[o.name] = o.default;
  });
  options.callbacks.forEach((o) => {
    defaults[o.name] = () => {};
  });

  return defaults;
};

export default options;
